### PR TITLE
Early allocation of the Chef::RunContext

### PR DESCRIPTION
**Pull request by lamont-granquist**
Allow for allocation of the Chef::RunContext without its arguments and
use that to allocate it early.  The APIs are kept backcompatible since
otherwise this has a decent chance of blowing up something like
chefspec.

### CODE REVIEW

File: lib/chef/data_bag.rb [hunk 1]
```diff
@@ -32,8 +32,8 @@ class DataBag
     include Chef::Mixin::FromFile
     include Chef::Mixin::ParamsValidate
 
-    VALID_NAME = /^[\.\-[:alnum:]_]+$/.freeze
-    RESERVED_NAMES = /^(node|role|environment|client)$/.freeze
+    VALID_NAME = /^[\.\-[:alnum:]_]+$/
+    RESERVED_NAMES = /^(node|role|environment|client)$/
```
Comment (lamont-granquist) [inline lib/chef/data_bag.rb]: unrelated chefstyle that somehow crept in...

**Comment by jaymzh** [comment]
Is the idea of this so that we can log earlier and have earlier event handlers?

File: lib/chef/client.rb [hunk 2]
```diff
@@ -451,9 +473,9 @@ def sync_cookbooks
     #
     # @api private
     def setup_run_context
-      run_context = policy_builder.setup_run_context(specific_recipes)
+      @run_context = policy_builder.setup_run_context(specific_recipes, run_context)
       assert_cookbook_path_not_empty(run_context)
-      run_status.run_context = run_context
+      run_status.run_context = run_context # backcompat for chefspec
```
Comment (jaymzh) [inline lib/chef/client.rb]: should this be @run_context since you just assigned to that?

Comment (lamont-granquist) [reply lib/chef/client.rb]: no, its an accessor so it works fine and returns `@run_context`

**Comment by lamont-granquist** [comment]
it ultimately gets the run_context created earlier and gets setup_run_context out of the business of creating it, and that allows event handlers to get ahold of that object earlier
and then the point is more to the hang the rest client(s) off of the run_context for reuse.  that has broader benefits, but one goal is to get all the ResourceReporter + DataCollector crap out of the Chef::Client.

**Comment by lamont-granquist** [comment]
also "do less work in initializers" and now you can create a Chef::RunContext without passing a node + cookbook_collection + event handler, which it doesn't always need (noting how often we pass `{}` in for the cookbook_collection), etc.

not that i'm certain how much i want to push that, since there's about a billion run_context objects created in the unit tests and i don't really want to go through and 'fix' every single one just for philosophical reasons -- hence why this change is all backcompat and not breaking...

**Pull request merged**